### PR TITLE
Production release scripts [not ready for merging]

### PR DESCRIPTION
**Pull request by pde**
This branch contains the release script we used for 0.1.0, plus some improvements.

Closes: #1185.

### CODE REVIEW

**Comment by pde** [comment]
At the moment, I'm seeing some failures in the tests that this script runs:
```
2015-12-11 17:59:20,039:DEBUG:letsencrypt.reverter:Creating backup of /tmp/tmpRvYhIitemp/etc_nginx/sites-enabled/example.com
2015-12-11 17:59:20,039:DEBUG:letsencrypt.reverter:Creating backup of /tmp/tmpRvYhIitemp/etc_nginx/sites-enabled/example.com
2015-12-11 17:59:20,039:DEBUG:letsencrypt.reverter:Creating backup of /tmp/tmpRvYhIitemp/etc_nginx/sites-enabled/example.com
2015-12-11 17:59:20,041:DEBUG:letsencrypt.reverter:Creating backup of /tmp/tmpRvYhIitemp/etc_nginx/server.conf
2015-12-11 17:59:20,041:DEBUG:letsencrypt.reverter:Creating backup of /tmp/tmpRvYhIitemp/etc_nginx/server.conf
2015-12-11 17:59:20,041:DEBUG:letsencrypt.reverter:Creating backup of /tmp/tmpRvYhIitemp/etc_nginx/server.conf
...........
======================================================================
ERROR: Failure: ImportError (cannot import name errors)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/acme/acme/jose/json_util_test.py", line 10, in <module>
    from acme.jose import errors
ImportError: cannot import name errors

======================================================================
ERROR: Failure: ImportError (cannot import name errors)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/acme/acme/jose/jwa_test.py", line 6, in <module>
    from acme.jose import errors
ImportError: cannot import name errors

======================================================================
ERROR: Failure: ImportError (cannot import name errors)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/acme/acme/jose/jwk_test.py", line 7, in <module>
    from acme.jose import errors
ImportError: cannot import name errors

======================================================================
ERROR: Failure: ImportError (cannot import name errors)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/acme/acme/jose/jws_test.py", line 10, in <module>
    from acme.jose import errors
ImportError: cannot import name errors

======================================================================
ERROR: Failure: ImportError (cannot import name errors)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/acme/acme/challenges_test.py", line 10, in <module>
    from acme import errors
ImportError: cannot import name errors

======================================================================
ERROR: Failure: ImportError (cannot import name challenges)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/acme/acme/client_test.py", line 11, in <module>
    from acme import challenges
ImportError: cannot import name challenges

======================================================================
ERROR: Failure: ImportError (cannot import name errors)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/acme/acme/crypto_util_test.py", line 9, in <module>
    from acme import errors
ImportError: cannot import name errors

======================================================================
ERROR: Failure: ImportError (cannot import name challenges)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/acme/acme/messages_test.py", line 6, in <module>
    from acme import challenges
ImportError: cannot import name challenges

======================================================================
ERROR: Failure: ImportError (cannot import name challenges)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/venv/local/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/acme/acme/standalone_test.py", line 14, in <module>
    from acme import challenges
ImportError: cannot import name challenges

======================================================================
FAIL: test_json_dumps_pretty (acme.jose.interfaces_test.JSONDeSerializableTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/pde/eff/chocolate/chocolate/letsencrypt/releases/le.0.2.0.dev201512111.14217/acme/acme/jose/interfaces_test.py", line 94, in test_json_dumps_pretty
    '[\n    "foo1",\n    "foo2"\n]')
AssertionError: '[\n    "foo1", \n    "foo2"\n]' != '[\n    "foo1",\n    "foo2"\n]'

----------------------------------------------------------------------
Ran 702 tests in 30.867s

FAILED (errors=9, failures=1)
```

**Comment by pde** [comment]
Fortunately, this does not need to be merged tonight.  Unlike changes to the client itself, the release script does not require or benefit from developer canary time.

**Comment by pde** [comment]
If anyone is playing with it though, make a test GPG key and set RELEASE_GPG_KEY to its id.

File: letsencrypt/cli.py [hunk 1]
```diff
@@ -889,7 +889,7 @@ def prepare_and_parse_args(plugins, args):
         version="%(prog)s {0}".format(letsencrypt.__version__),
         help="show program's version number and exit")
     helpful.add(
-        "automation", "--renew-by-default", action="store_true",
+        "automation", "--renew-by-default", "--replace", action="store_true",
```
Comment (bmw) [inline letsencrypt/cli.py]: Why is this change in this PR?

File: tools/dev-release.sh [hunk 2]
```diff
@@ -1,13 +1,43 @@
-#!/bin/sh -xe
+#!/bin/bash -xe
 # Release dev packages to PyPI
 
+Usage() {
+    echo Usage:
+    echo "$0 [ --production ]"
```
Comment (bmw) [inline tools/dev-release.sh]: nit: If I'm reading the code correctly, if you use `--production` you also need to include the current release version and the next version? This should be reflected here as I wouldn't have known that unless I read the code.

File: tools/dev-release.sh [hunk 3]
```diff
@@ -1,13 +1,43 @@
...
+    fi
+}
+
+if [ "$1" = "--production" ] ; then
+    version="$2"
+    CheckVersion Version "$version"
+    echo Releasing production version "$version"...
+    nextversion="$3"
+    CheckVersion "Next version" "$nextversion"
+    RELEASE_BRANCH="master"
```
Comment (bmw) [inline tools/dev-release.sh]: Given we're building releases off of a few day old version of `master`, is this the branch we want to use?

**Comment by bmw** [comment]
Testing this script, I'm getting the same errors as you did earlier. Have you made any progress on identifying/fixing the problem? If not, I can look into if you'd like.

**Pull request merged**